### ISSUE TITLE
API authentication workflow for apps

### ISSUE THREAD

**Issue by lwojcik** [jointwt/twtxt#139]
**Is your feature request related to a problem? Please describe.**
I've seen #126 and I believe it's a major step forward compared to the current state of affairs.

However, if I understood it correctly, API tokens are still tied directly to user profile. It leads to a chicken and egg problem: I won't register a new account via the API unless I already have a working account to work with. This is fine for personal projects like [automating my Dilbert feed](https://twtxt.net/user/dilbert) that aren't designed to scale but very suboptimal for external tools meant to be used by the community.

Also, API authentication workflow is something I'll have to explain in the docs for my JS wrapper and this will be tricky to get right. :) 

**Describe the solution you'd like**

App-based API authentication workflow that would make it easier for third-party projects to communicate with twt pods in a secure manner, e.g. identifying themselves as _apps_, not _users_.

Let me create an app in my profile settings and provide me with a pair of automatically generated credentials: client ID and client secret. Require me to pass them to the `auth` endpoint in order to get a valid token. For security reasons I should be able to regenerate client secret any time I decide. If API ever implements usage limits, tying them to registered apps will make it easier to enforce them.

**Describe alternatives you've considered**

From the perspective of a potential third-party app developer, there are currently three sane workarounds:

1. Using an empty profile just for the sake of API authentication on a selected pod. If this profile gets accidentally deleted, my app breaks. If someone hacks their way into this profile, my app is compromised. If I lose access to the account, my app is a zombie until I create a new acount and update my app (c'mon Google, approve my app quicker or my users downvote it to oblivion!). And that still implies I was smart enough to actually create a separate profile for all of this.

2. Redirecting user to [register form](https://twtxt.net/register) or embedding this form within the app (e.g. web view in a mobile app) so that user doesn't have to switch contexts. Still creates friction as I can't design onboarding experience the way I want.

3. Or just giving up and limiting core app functionality to existing users only. No new users allowed.

**Additional context**
N/A

**Comment by prologic** [comment]
What you are effectively describing is basically OAuth right? 😀

I _think_ you missed something perhaps non-obvious... You _can_ currently register a new account via the API, the endpoints that effectively require no prior authentication/credentials are:

- /api/v1/ping
- /api/v1/register
- /api/v1/discover

However as per my OAuth comment, I _believe_ you are right, we really should also implement App tokens. I just hadn't thought this all the way through and put this API together quite quickly for @aaadonai and @dooven 

Does someone want to take this up? Do we need to refactor the current API or have an `/api/v2/`?

**Comment by lwojcik** [comment]
> What you are effectively describing is basically OAuth right? grinning

Yeah. For a full OAuth workflow we'd also need endpoints for validating and refreshing token, but in essence that's what I meant.

> You can currently register a new account via the API, the endpoints that effectively require no prior authentication/credentials are:

Yup, just confirmed it was my bad. I had a bit of hard time figuring this out based on [API docs](http://dev.twtxt.net/doc/api.html) and error message '400 Bad request' was a bit vague. I'll see if I can improve API docs to remove that kind of confusion later.

**Comment by prologic** [comment]
Thanks @lwojcik ! Also please feel free to work with @dooven to improve the API.
